backend/sirius-components-diagrams-layout/src/main/java/org/eclipse/sirius/components/diagrams/layout/incremental/provider/NodePositionProvider.java

```diff
@@ -51,116 +54,160 @@
      */
     public Position getPosition(Optional<IDiagramEvent> optionalDiagramElementEvent, NodeLayoutData node) {
         Position position = node.getPosition();
+        if (NodeType.NODE_LIST_ITEM.equals(node.getNodeType())) {
+            Optional<Position> nodeListItemPosition = this.getNodeListItemPosition(node);
+            if (nodeListItemPosition.isPresent()) {
+                position = nodeListItemPosition.get();
+            }
+        } else if (optionalDiagramElementEvent.isPresent()) {
+            IDiagramEvent diagramEvent = optionalDiagramElementEvent.get();
+            Optional<Position> eventRelativePosition = this.getEventRelativePosition(diagramEvent, node);
+            if (eventRelativePosition.isPresent()) {
+                position = eventRelativePosition.get();
+            }
+        }
+        return position;
+    }
 
-        Optional<Position> optionalPosition = this.getSpecificNodePositionFromEvent(optionalDiagramElementEvent, node);
-        if (optionalPosition.isPresent() && !NodeType.NODE_LIST_ITEM.equals(node.getNodeType())) {
-            position = optionalPosition.get();
-        } else if (!this.isAlreadyPositioned(node) || NodeType.NODE_LIST_ITEM.equals(node.getNodeType())) {
-            Optional<Position> optionalStartingPosition = this.getOptionalStartingPositionFromEvent(optionalDiagramElementEvent);
-            if (optionalStartingPosition.isPresent() && this.last == null && !NodeType.NODE_LIST_ITEM.equals(node.getNodeType())) {
-                Position parentPosition = node.getParent().getAbsolutePosition();
-                double eventX = optionalStartingPosition.get().getX();
-                double eventY = optionalStartingPosition.get().getY();
-                if (this.isEventPositionInParentBounds(node.getParent(), Position.at(eventX, eventY))) {
-                    // The node has been created by a tool and has a fixed position
-                    double xPosition = eventX - parentPosition.getX();
-                    double yPosition = eventY - parentPosition.getY();
-                    position = Position.at(xPosition, yPosition);
-                    this.last = node;
-                } else {
-                    // The new node should appear in the parent, without moving the parent unless necessary
-                    position = Position.at(10, 10);
-                    this.last = node;
```

Comment:
The main interesting thing here is that I create a special case for the NodeType.NODE_LIST_ITEM nodes, instead of having them mingled with the rest. I believe it helps to understand what the algorithm does.